Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,392 @@
+    //  - etc.
+    //  These can be done using the DWrite IDWriteTextFormat range property methods.
+    NSRange curRange = NSMakeRangeFromCF(range);
+    /*
+    NSDictionary* attribs = [ts->_attributedString attributesAtIndex:0 effectiveRange:&curRange];
+    UIFont* font = [attribs objectForKey:static_cast<NSString*>(kCTFontAttributeName)];
+    if (font == nil) {
+        font = [s_lazyUIFont systemFontOfSize:c_defaultSystemFontSize];
+    }
+    float fontSize = [font pointSize];
+    */
+
+    wchar_t* wcharString = reinterpret_cast<wchar_t*>(
+        const_cast<char*>([[ts->_string substringWithRange:curRange] cStringUsingEncoding:NSUTF16StringEncoding]));
+
+    // TODO::
+    // We need too support widthFunc semantic to be able to support NSLayout*. We could either change the API signature of this API or
+    // create a new API to support dynamic width support.
+    //
+    // float width = widthFunc(widthParam, 0, 0, fontSize);
+
+    // Get the direct write factory instance
+    ComPtr<IDWriteFactory> dwriteFactory = GetDWriteFactoryInstance();
+
+    ComPtr<IDWriteTextLayout> textLayout;
+    THROW_IF_FAILED(dwriteFactory->CreateTextLayout(
+        wcharString, ts->_charactersLen, textFormat.Get(), frameSize.size.width, frameSize.size.height, &textLayout));
+
+    return textLayout;
+}
+
+/**
+ * Custom IDWriteTextRenderer class that captures the glyph runs generated by DWrite for the given TextLayout and TextRenderer constraints.
+ */
+class CustomTargetTextRenderer : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteTextRenderer> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    CustomTargetTextRenderer();
+
+    HRESULT RuntimeClassInitialize();
+
+    HRESULT STDMETHODCALLTYPE DrawGlyphRun(_In_ void* clientDrawingContext,
+                                           _In_ float baselineOriginX,
+                                           _In_ float baselineOriginY,
+                                           DWRITE_MEASURING_MODE measuringMode,
+                                           _In_ DWRITE_GLYPH_RUN const* glyphRun,
+                                           _In_ DWRITE_GLYPH_RUN_DESCRIPTION const* glyphRunDescription,
+                                           _In_ IUnknown* clientDrawingEffects) throw() {
+        _DWriteGlyphRunDetails* glyphs = static_cast<_DWriteGlyphRunDetails*>(clientDrawingContext);
+        glyphs->_baselineOriginX.push_back(baselineOriginX);
+        glyphs->_baselineOriginY.push_back(baselineOriginY);
+
+        _DWriteGlyphRun glyphRunInfo;
+        glyphRunInfo._glyphCount = glyphRun->glyphCount;
+        for (int index = 0; index < glyphRun->glyphCount; index++) {
+            glyphRunInfo._glyphIndices.push_back(glyphRun->glyphIndices[index]);
+            glyphRunInfo._glyphAdvances.push_back(CGSizeMake(glyphRun->glyphAdvances[index], 0));
+        }
+        glyphs->_glyphRuns.push_back(glyphRunInfo);
```

Comment:
Synced offline - as this code is going away in the next iteration leaving it as is

---
In reply to: [78079893](https://github.com/Microsoft/WinObjC/pull/950#discussion_r78079893) [](ancestors = 78079893,77864667,77756323)